README.md
<p align="center">
    <img src="https://user-images.githubusercontent.com/6702424/76674598-91ebfc00-65b1-11ea-88df-eb43f04f3cce.png">
</p>
<p align="center">
    💧<i>EventEmitter's typesafe replacement</i>💧
    <br>
    <br>
    <img src="https://github.com/garronej/evt/workflows/ci/badge.svg?branch=develop">
    <img src="https://img.shields.io/bundlephobia/minzip/evt">
    <img src="https://img.shields.io/npm/dw/evt">
    <img src="https://img.shields.io/npm/l/evt">
</p>

</p>
<p align="center">
  <a href="https://www.evt.land">Home</a>
  -
  <a href="https://docs.evt.land/overview">Documentation</a>
  -
  <a href="https://github.com/garronej/evt/pull/16">v2.0 🚀</a>
</p>

---

`'evt'` is intended to be a replacement for `'events'`.  
It enables and encourages **functional programming** and makes heavy use of **typescript**'s type inference features to provide **type safety** while keeping things **concise and elegant** 🍸.

<b>Suitable for any JS runtime env (deno, node, old browsers, react-native ...)</b>
- ✅ It is both a [Deno](https://deno.land/x/evt) and an [NPM](https://www.npmjs.com/evt) module. ( Achieved with [Denoify](https://github.com/garronej/denoify) ) 
- ✅ Lightweight, no dependency.
- ✅ Can be imported with `require` (CJS) or `import` (ESM) as long as you install [`evt@beta`](https://github.com/garronej/evt/pull/16)
- ✅ [React Hooks integration](https://stackblitz.com/edit/evt-react-hooks-todo-list?file=index.tsx)

Can be imported in TypeScript projects using version &gt;= **3.4** \(Mar 2019\) and in any plain JS projects.

# TL;DR*

```typescript
import { Evt } from "evt";

const evtText = new Evt<string>();
const evtTime = new Evt<number>();

evtText.attach(text => console.log(text));
evtTime.attachOnce(time => console.log(time));

evtText.post("hi!"); //Prints "hi!"
evtTime.post(123);   //Prints "123"
evtTime.post(1234);  //Prints nothing
```
OR
```typescript
import { Evt, to } from "evt";

const evt = new Evt<
    [ "text",  string ] | 
    [ "time",  number ]
>();

//Mind the '$' prefixing 'attach'
evt.$attach(to("text"), text => console.log(text));
evt.$attachOnce(to("time"), time => console.log(time));

evt.post(["text", "hi!"]);
evt.post(["time", 123]);
evt.post(["time", 1234]);
```
in React, it let you attach event listeners without having to worry about detaching them.
```typescript
import { useState } from "react";
import { Evt } from "evt";
import { useEvt } from "evt/hooks";

const evtTick = Evt.create();

setInterval(()=> evtTick.post(), 1000);

function App(){

    const [count, setCount]= useState(0);

    useEvt(ctx=> {
    
        evtTick.attach(ctx, ()=> setCount(count+1));
    
    },[count]);
    
    return <h1>tick count: {count}</h1>;


}
```
[run it](https://stackblitz.com/edit/evt-hooks-101?file=index.tsx)

_*Those are introductory examples, EVT can do much more than this._

# Who is using it

<p align="center">
    <a href="https://connext.network">
        <img src="https://user-images.githubusercontent.com/6702424/84102640-4e1e5c80-aa11-11ea-9d13-df0a65c8cdaf.png">
    </a>
    <a href="https://www.semasim.com">
        <img src="https://user-images.githubusercontent.com/6702424/84102785-aead9980-aa11-11ea-915b-5c4a5282c44e.png"> 
    </a>
    <br>
    <a href="https://thegraph.com">
        <img src="https://user-images.githubusercontent.com/6702424/85961024-ed4aca00-b9a7-11ea-9c7e-03e2f59e6d21.png"> 
    </a>
    <a href="https://insee.fr">
        <img src="https://user-images.githubusercontent.com/6702424/117936881-a9358f00-b305-11eb-84b9-e61593632bdd.png"> 
    </a>
</p>

# Install / Import

## In Deno:
```typescript
import { Evt } from "https://deno.land/x/evt/mod.ts";
```
## Anywhere else:
```bash
$ npm install --save evt
# OR npm install --save evt@beta for the latest features
```
```typescript
import { Evt } from "evt"; 
```

## Import from HTML, with CDN

```html
<script src="//unpkg.com/evt/bundle.min.js"></script>
<script>
    const { Evt } = window["evt"];
</script>
```

# Try it

[Run some examples](https://stackblitz.com/edit/evt-playground?embed=1&file=index.ts&hideExplorer=1)

<p align="center"> 
    <img src="https://www.evt.land/assets/img/try-in-browser.gif">  
</p>

# Motivations

There are a lot of things that can't easily be done with `EventEmitter`:

* Enforcing **type safety**.
* Removing a particular listener ( if the callback is an anonymous function ).
* Adding a one-time listener for the next event that meets a condition.
* Waiting \(via a Promise\) for one thing or another to happen.  
_Example: waiting at most one second for the next message, stop waiting if the socket disconnects._

Why would someone pick EVT over RxJS:  

* RxJS introduces a lot of abstractions. It's a big jump from ``EventEmitter``.
* With RxJS It is often needed to resort to custom [type guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards), the filter operator [breaks the type inference.](https://stackblitz.com/edit/evt-795plc?embed=1&file=index.ts&hideExplorer=1)
* RxJS tends to be quite verbose.
* It could be months before RxJS it eventually supports Deno.
* No official guideline on how to integrate RxJS with React.

EVT is an attempt to address all these points while trying to remain as accessible as `EventEmitter`.  
  
</br>

[Get started](https://docs.evt.land/overview#rxjs-comparison)

# The sticker

<p align="center">
    <img src="https://user-images.githubusercontent.com/6702424/90626180-e3607d00-e21a-11ea-9c88-82880ac9cedf.png">  
</p>
<p align="center">
    <a href="https://teespring.com/fr/evt-sticker">Shop</a>
</p>


CHANGELOG.md
### **1.11.2** (2022-06-01)  
  
- Fix useRerenderOnStateChange    
  
### **1.11.1** (2022-06-01)  
  
- Correct error in useEvt    
  
## **1.11.0** (2022-06-01)  
  
- hooks compat React 18  
- Create init-onyxia.sh  
- Merge pull request #24 from garronej/dependabot/npm_and_yarn/shelljs-0.8.5

Bump shelljs from 0.8.4 to 0.8.5    
  
### **1.10.2** (2021-12-23)  
  
- Fix for Deno 1.17  
- Update README.md  
- Update README.md  
- Update README.md    
  
### **1.10.1** (2021-07-20)  
  
-  #21: Feat distinct operator    
  
## **1.10.0** (2021-07-20)  
  
- #21: Feat distinct operator    
  
### **1.9.14** (2021-05-08)  
  
- Deno fix: tools/inDepth --no-check compat    
  
### **1.9.13** (2021-04-16)  
  
- #19: Fix compat with Deno v1.9    
  
### **1.9.12** (2020-12-29)  
  
- Support --no-check deno flag  
- Add React example in the readme    
  
### **1.9.11** (2020-12-05)  
  
- Indicate that v2 is under devloppement.    
  
### **1.9.10** (2020-11-25)  
  
- Fix issue related to strict mode hooks/useEvt    
  
### **1.9.9** (2020-11-16)  
  
- Create publish-beta.yaml    
  
### **1.9.8** (2020-11-16)  
  
- Add nonNullable operator    
  
### **1.9.7** (2020-11-15)  
  
- Add tools/typeSafety/PromiseOrNot helper type    
  
### **1.9.6** (2020-11-15)  
  
- Improve Evt.asyncPipe()    
  
### **1.9.5** (2020-11-14)  
  
- Ensure semantic guarantee for hooks/useEvt    
  
### **1.9.4** (2020-11-10)  
  
- Add the Id helper type in tools/typeSafety/id    
  
### **1.9.3** (2020-11-09)  
  
- Add tools/typeSafety/MethodNames    
  
### **1.9.2** (2020-10-29)  
  
- Improve error stack trace of tools/typeSafety/assert    
  
### **1.9.1** (2020-10-28)  
  
- Make EventEmitter.removeListener easily emulable    
  
## **1.9.0** (2020-10-24)  
  
- Add evt.postAndWait()    
  
### **1.8.11** (2020-10-22)  
  
- Fix TypeScript error in Evt.asNonPostable.d.ts '=' expected. TS1005
(create-react-app) Use of import type was breaking the TypeScript retroconpatibility.    
  
### **1.8.10** (2020-09-14)  
  
- Merge pull request #14 from garronej/issue-12-support-strict-type-import-export

Issue 12 support strict type import export ( fix for retrocompat )  
- Use --unstable flag to run tests  
- CI workflow runs test on Windows as well  
- fix: missed some
fix: add type imports to test files  
- fix: use import/export type syntax for types  
- Merge pull request #11 from garronej/dependabot/npm_and_yarn/node-fetch-2.6.1

Bump node-fetch from 2.6.0 to 2.6.1  
- Update README.md  
- Update README.md    
  
### **1.8.9** (2020-09-03)  
  
- add Evt.asyncPipe (as a static method until async operator lands)    
  
### **1.8.8** (2020-08-28)  
  
- Hooks denoified as well  
- Update README.md    
  
### **1.8.6** (2020-08-08)  
  
- Advertise React Hooks integration    
  
### **1.8.5** (2020-08-07)  
  
- Merge pull request #8 from garronej/dependabot/npm_and_yarn/lodash-4.17.19

Bump lodash from 4.17.15 to 4.17.19  
- Fix issue #9: Bug with Evt.from(window,...)    
  
### **1.8.4** (2020-08-05)  
  
- Switch back to tag name prefixed with 'v'  
- fix useEvt hook for verry slow browser    
  
### **1.8.3** (2020-07-31)  
  
- Add documentation for react hooks    
  
### **1.8.2** (2020-07-31)  
  
- Fix for useEvt on Safari    
  
### **1.8.1** (2020-07-28)  
  
- fix bug in hooks/useStatefulEvt  
- Add missing 'attachExtract' overloads    
  
## **1.8.0** (2020-07-24)  
  
- Add hooks for react integration    
  
### **1.7.14** (2020-07-19)  
  
- update dependencies    
  
### **1.7.13** (2020-06-14)  
  
- Add JSDoc link to Evt.merge  
- update deps  
- Update project description  
- Reference Denoify in the readme  
- include who is using it sectiion    
  
### **1.7.12** (2020-05-30)  
  
- Reference Denoify in the readme  
- Update README.md    
  
### **1.7.11** (2020-05-25)  
  
  
  
### **1.7.11** (2020-05-25)  
  
- Re include source map without breaking TS retro compatibility  
  
### **1.7.10** (2020-05-25)  
  
- Fix: Include CDN build    
  
### **1.7.9** (2020-05-24)  
  
- Do not include source in the NPM bundle for TS retro compatibility (they will be re-included later)
  
### **1.7.8** (2020-05-24)  
  
- Remove nullish coalescing from source for stackblitz compat  
  
### **1.7.7** (2020-05-24)  
  
- Not including source test files in the NPM bundle  
- readme typo fix    
  
### **1.7.6** (2020-05-23)  
  
- Make the README links visible on deno.land/x  
- Merge pull request #7 from searchableguy/develop
- Add a note in overview about difference between attach and $attach  
- Update run-exclusive    
  
### **1.7.5** (2020-05-23)  
  
- Bump version ( changelog ignore )  
- Use test/mod.ts instead of test/index_deno.ts to align with conventions  
- Enable source map, 'Go to Definition' navigate to .ts source file instead of .d.ts file  
- Revert back to the old type import syntax and regain support for TS 3.4 - 3.7, WARNING: This release do not work with Deno 1.0.1 but do work with 1.0.2  
- Typo fixes
  
### **1.7.4** (2020-05-22)  
  
- GitBook: [develop] one page modified  
- instruct to refer to CHANGELOG.md  
- GitBook: [develop] one page modified  
- GitBook: [develop] 32 pages and one asset modified    
  
### **1.7.3** (2020-05-21)  
  
- fix for supporting Deno v1.0.1, temporary dropping support for TS < 3.8    
  
### **1.7.2** (2020-05-21)  
  
- Notice about the fact EVT won't work on Deno 1.0.1  
- GitBook: [develop] 17 pages and one asset modified    
  
### **1.7.1** (2020-05-21)  
  
- Temporarily disabling sourcemaps  
- Add CI badge    
  

### **1.7.0** (2020-05-21)  
  
- Includes CDN build to import from HTML.
- Enable short path ( Breaking change ) ``evt/dist/tools/typeSafety`` => ``evt/tools/typeSafety``


src/readme.ts


/*

#ts-events-extended

A library intended to be a replacement for node's EventEmitter
featuring type safety and making use promises.

Similar to Qt signal/slot or C# events. 

Target es3, will run anywhere including in the browser ( with browserify ).
Will transpile used in projects with old version of typescript ( 2.8 and up ).

#History

This project was originally a fork aimed to add features to "ts-events".
Along way it has been re-implemented from scratch keeping only the 
core design philosophy of it's parent. 
AsyncEvent and QueuedEvent have been scraped out focusing only on the SyncEvent class.


*/


import { Evt } from "./lib";

{

    const evtText = new Evt<string>();
    //Unlike in node's events we use a different instance of SyncEvent
    //for every event type
    const evtTime = new Evt<number>();

    evtText.attach(text => console.log(text));
    evtTime.attachOnce(time => console.log(time));

    evtText.post("hi!");
    // at this point, "hi!" have been printed to the console.

    evtTime.post(123);
    // at this point, "123" have been printed to the console.

    evtTime.post(1234);
    //Nothing was printed to the console. ( attachOnce )

}


/*
Exact equivalent with node's EventEmitter:
*/

import { EventEmitter } from "events";

{

    const eventEmitter = new EventEmitter();

    eventEmitter.on("text", text => console.log(text));
    eventEmitter.on("time", time => console.log(time));

    eventEmitter.emit("text", "hi!");
    eventEmitter.emit("time", 123);


}

/*
# Uses of Promise, waiting until the next event is posted.
*/

{

    const evtText = new Evt<string>();

    (async ()=>{

        //waitFor return a promise that will resolve next time 
        //post() is invoked on evtText.
        const text = await evtText.waitFor();

        console.log(text);

    })();

    evtText.post("Hi");

}

//It is possible to set how long we wait for the next event before
//the promise returned by waitFor reject.

import {  EvtError } from "./lib";

{

    const evtText = new Evt<string>();


    (async ()=>{


        try{
        
            const text = await evtText.waitFor(500);

            console.log(text);

        }catch(error){

            console.assert(error instanceof EvtError.Timeout);
            //Error can be of type EvtError.Detached if the handler
            //was detached before the promise returned by waitFor have resolved.

            console.log("TIMEOUT!");

        }


    })();

    //A random integer between 0 and 1000
    const timeout= ~~(Math.random() * 1000);

    //There is a fifty-fifty chance "Hi!" is printed to the console
    //else "TIMEOUT!" is printed.
    setTimeout(
        ()=> evtText.post("Hi!"), 
        timeout
    );


}


//Filtering events.

/*
Matcher function are used to attach a callback that will handle 
only eventData that satisfies certain properties:
*/

{

    const evtText= new Evt<string>();

    evtText.attach(
        text=> text.startsWith("H"), //A matcher function take an argument of type T ( here string ) and returns a boolean.
        text=> {
            console.assert( text.startsWith("H") );
            console.log(text);
        }
    );

    //Nothing will be printed to the console.
    evtText.post("Bonjour");

    //"Hi!" will be printed to the console.
    evtText.post("Hi!");

}

/* Using type guard function as matcher function */

type Circle = {
    type: "CIRCLE";
    radius: number;
}


type Square = {
    type: "SQUARE";
    sideLength: number;
}

type Shape = Circle | Square;

const matchCircle = (shape: Shape): shape is Circle =>
    shape.type === "CIRCLE"
    ;

{


    const evtShape = new Evt<Shape>();

    evtShape.attach(
        matchCircle,
        shape => {

            //When the matcher function passed to attach is a Type Guard
            //the type of the event data will be inferred.
            //Here shape is of type Circle thus the TSC does not complain if we try to access 
            //the radius property.
            console.log(shape.radius);


        }
    );

    //Nothing will be printed to the console.
    evtShape.post({
        "type": "SQUARE",
        "sideLength": 3
    });

    //"33" Will be printed to the console.
    evtShape.post({
        "type": "CIRCLE",
        "radius": 33
    });

}

/* Combining Matcher and waitFor, attachOnce */

{

    const evtShape = new Evt<Shape>();

    evtShape.attachOnce(
        (shape): shape is Square => (
            shape.type === "SQUARE" &&
            shape.sideLength > 20
        ),
        ({ sideLength })=> console.log(`length: ${sideLength}`)
    );

    evtShape.waitFor(matchCircle)
        .then(circle => console.log(`radius: ${circle.radius}`))
        ;

    
    const circle: Circle = {
        "type": "CIRCLE",
        "radius": 33
    };

    //"radius: 33" will be printed to the console.
    evtShape.post(circle);

    //Nothing will be printed to the console, the promise returned by waitFor has already resolved.
    evtShape.post(circle);

    //Nothing will be printed, the side length is too short
    evtShape.post({
        "type": "SQUARE",
        "sideLength": 12
    });

    evtShape.post({
        "type": "SQUARE",
        "sideLength": 21
    });
    //"length: 21" have been  printed to the console.

    //Noting will be printed, attachOnce's callback function have already been invoked.
    evtShape.post({
        "type": "SQUARE",
        "sideLength": 44
    });

}



/*
#No arguments
A TypeScript annoyance: when you create an event with a void argument, TypeScript forces you to pass 'undefined' to post(). 
To overcome this, we added VoidSyncEvent class.
*/

import { VoidEvt } from "./lib";


{

    const evtSocketConnect = new VoidEvt();

    evtSocketConnect.attach(() => console.log("SOCKET CONNECTED"));


    evtSocketConnect.post();
    //"SOCKET CONNECTED" have been printed to the console.


}



//Handler priority.
//The handlers callback functions are invoked in the order they have been attached
//unless attachPrepend is used.

{

    const evtConnect = new VoidEvt();

    evtConnect.attach(() => console.log("B"));
    evtConnect.attach(() => console.log("C"));

    evtConnect.attachPrepend(() => console.log("A"));

    evtConnect.post();
    //"A", "B", "C" is printed to the console.

}

//Extracting events.
//To handle hedge cases that haven't been anticipated without having to rethink the all model
//we provide a way to extract particular event.


{

    const evtCircle = new Evt<Circle>();


    evtCircle.attach(
        circle => {
            console.assert(circle.radius > 0);
        }
    );

    evtCircle.attachExtract(
        ({ radius }) => radius <= 0,
        ({ radius }) => console.log(`malformed circle with radius: ${radius} extracted`)
    );

}

//Detaching events

{

    const evtText = new Evt<string>();

    //detach with no argument will detach all handlers (attach, attachOnce, waitFor... )
    evtText.detach();

}




//To detach a particular handler for which we have the reference of the callback function:
{

    const evtText = new Evt<string>();

    const callback = (_text: string) => { };

    evtText.attach(callback);

    //evtText.getHandlers().find(handler => handler.callback === callback)?.detach();
    const h = evtText.getHandlers().find(handler => handler.callback === callback);

    if( h ){
        h.detach();
    }

}

//By far the preferred way of detaching an handler is by using "boundTo" context 
//as more often that not we don't keep the reference of the callback function.
{

    const evtText = new Evt<string>();

    //boundTo can be anything but a number, a callable function (i.e. not a constructor), undefined  or null.
    const boundTo = [];

    evtText.attach(
        boundTo,
        _text => { }
    );

    evtText.attachOnce(
        boundTo,
        _text => { }
    );

    evtText.detach(boundTo);

}

//A more advanced example here detaching all handler that have a given matcher:
{

    const evtShape = new Evt<Shape>();

    evtShape.attach(
        matchCircle,
        _circle => { }
    );
    evtShape.attachOnce(
        matchCircle,
        _circle => { }
    );

    evtShape.waitFor(matchCircle)
        .then(_circle => { })
        ;

    //waitFor will no reject once detached as no timeout have been specified.
    evtShape.getHandlers()
        .filter(({ matcher }) => matcher === matchCircle)
        .forEach(({ detach }) => detach())
        ;

}

//Misc
{

    const evtText = new Evt<string>();

    //Number of type post() have been called.
    const n = evtText.postCount;

    console.assert(n === 0);

    //A SyncEvent<Handler<string>> that track when handler are attached to evt.
    evtText.evtAttach;

}


// Combining Once, Prepend, matcher, timeout and boundTo

/*

Large number of methods combining Once, Prepend are exposed.

![Screenshot 2020-02-06 at 21 43 27](https://user-images.githubusercontent.com/6702424/73977452-b3125a80-492a-11ea-9657-4434a4c21d29.png)


For each of those methods a large number of overload are defined
so that you can combine matcher, timeout or boundTo.

![Screenshot 2020-02-06 at 21 43 59](https://user-images.githubusercontent.com/6702424/73977365-865e4300-492a-11ea-8e51-c304ade3fa41.png)


*/

//When we are not sure if the handlers have been attached already
//postOnceMatched can be used.

{


    const evtText = new Evt<string>();

    evtText.postOnceMatched("Foo Bar");

    //"before"\n"Foo Bar" will be printed to the console.
    evtText.attachOnce(text=> console.log(text));

    console.log("before");

}

//ObservableImpl is a class that allow to track mutation on
//a particular data type.

import { Observable, ObservableImpl } from "./lib";

{

    const obsText= new ObservableImpl<string>("foo");


    console.assert(obsText.value === "foo");

    obsText.evtChange.attach(
        newValue=> {

            console.assert(newValue === obsText.value);

            console.log(`newValue: ${newValue}`);

        }
    );

    //Nothing will be printed to the console as the value did not change.
    obsText.onPotentialChange("foo");


    obsText.onPotentialChange("bar");
    //"newValue: bar" have been printed to the console.

    console.assert(obsText.value === "bar");


    //ObservableImpl is assignable to Observable but
    //Observable is missing the onPotentialChange method.
    //It is used to expose an observable that should not be
    //modified by the user.
    const exposedObsText: Observable<string> = obsText;

    exposedObsText.value;
    exposedObsText.evtChange;

}




































//Edge cases: 

{


    const evtText = new Evt<string>();

    (async ()=>{

        const text1= await evtText.waitFor();

        const text2 = await evtText.waitFor(); 

        console.log(`${text1} ${text2}`);


    })();

    evtText.post("FOO");
    evtText.post("BAR");

    //"FOO BAR" is printed to the console ( Voodoo involved )





}

















/*

To track changes of the data we use a library that we have build in house called "ts-event-extended"

node's EventEmitter was not designed to support types so we build our onw alternative.

*/








